lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp

```diff
@@ -117,13 +119,40 @@ static void diagnoseStaticReports(const SILInstruction *I,
 // SWIFT_ENABLE_TENSORFLOW
 /// \brief Emit a diagnostic for `poundAssert` builtins whose condition is
 /// false or whose condition cannot be evaluated.
-/// TODO(clattner / marcrasi): Implement this for real.
 static void diagnosePoundAssert(const SILInstruction *I, SILModule &M) {
-  if (auto *builtinInst = dyn_cast<BuiltinInst>(I)) {
-    if (builtinInst->getBuiltinKind() == BuiltinValueKind::PoundAssert) {
-      diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
-               diag::pound_assert_not_implemented);
-    }
+  auto *builtinInst = dyn_cast<BuiltinInst>(I);
+  if (!builtinInst ||
+      builtinInst->getBuiltinKind() != BuiltinValueKind::PoundAssert)
+    return;
+
+  // TODO(marcrasi / clattner): Instantiate this earlier so that its cache
+  // is useful. We don't right now, because instantiating one for every
+  // function greatly slows down or hangs compilation (e.g. compiling
+  // libswiftStdlibUnicodeUnittest.so slows down from <1min to >1hr).
+  ConstExprEvaluator constantEvaluator(M);
+
+  SmallVector<SymbolicValue, 1> values;
+  constantEvaluator.computeConstantValues({builtinInst->getArguments()[0]},
+                                          values);
+  SymbolicValue value = values[0];
+  if (!value.isConstant()) {
+    // TODO(marcrasi / clattner): Emit more informative error.
+    diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
+             diag::pound_assert_condition_not_constant);
+    return;
+  }
+  assert(value.getTypeKind() == SymbolicValue::TKInteger &&
+         "sema prevents non-integer #assert condition");
+
+  APInt intValue = value.getIntegerValue();
+  assert(intValue.getBitWidth() == 1 &&
+         "sema prevents non-int1 #assert condition");
+  if (intValue.isNullValue()) {
+    auto *message = dyn_cast<StringLiteralInst>(builtinInst->getArguments()[1]);
```

Comment:
Please use cast<> instead of dyn_cast<> + assert.

Reply:
done